File: tests/oracle/rangemap_common_refinement_1204.cc
/*
 * Locked Ghidra 12.0.4 rangemap<> oracle for
 * RANGEMAP-COMMON-REFINEMENT-0001.
 *
 * This fixture exercises the template directly.  The observable is the
 * record identity sequence produced by the public iterators, so duplicate
 * visits expose the common-refinement partitions without inspecting private
 * AddrRange state.
 */

#include "rangemap.hh"

#include <iostream>
#include <sstream>
#include <string>

namespace {

using namespace ghidra;
using std::string;

class TestRecord {
public:
  class Subsort {
    int value;
  public:
    Subsort(void) : value(0) {}
    explicit Subsort(int val) : value(val) {}
    explicit Subsort(bool latest) : value(latest ? 1000000 : 0) {}
    bool operator<(const Subsort &other) const { return value < other.value; }
  };

  struct InitData {
    string name;
    int subsort;
    InitData(const string &nm, int sub) : name(nm), subsort(sub) {}
  };

  typedef unsigned long long linetype;
  typedef Subsort subsorttype;
  typedef InitData inittype;

private:
  string name;
  linetype first;
  linetype last;
  Subsort subsort;

public:
  TestRecord(const InitData &data, linetype a, linetype b)
    : name(data.name), first(a), last(b), subsort(data.subsort) {}
  linetype getFirst(void) const { return first; }
  linetype getLast(void) const { return last; }
  Subsort getSubsort(void) const { return subsort; }
  const string &getName(void) const { return name; }
};

typedef rangemap<TestRecord> TestMap;

string names(TestMap::const_iterator iter, TestMap::const_iterator end)
{
  std::ostringstream out;
  bool first = true;
  while (iter != end) {
    if (!first) out << ',';
    first = false;
    out << (*iter).getName();
    ++iter;
  }
  return out.str();
}

string listNames(const TestMap &map)
{
  std::ostringstream out;
  bool first = true;
  for (std::list<TestRecord>::const_iterator iter = map.begin_list();
       iter != map.end_list(); ++iter) {
    if (!first) out << ',';
    first = false;
    out << iter->getName();
  }
  return out.str();
}

void dumpFind(const string &label, const TestMap &map,
              TestRecord::linetype point)
{
  std::pair<TestMap::const_iterator, TestMap::const_iterator> result =
      map.find(point);
  std::cout << "case=" << label << "|find=" << point
            << "|records=" << names(result.first, result.second) << '\n';
}

void dumpFindSubsort(const string &label, const TestMap &map,
                     TestRecord::linetype point, int low, int high)
{
  std::pair<TestMap::const_iterator, TestMap::const_iterator> result =
      map.find(point, TestRecord::Subsort(low), TestRecord::Subsort(high));
  std::cout << "case=" << label << "|find_sub=" << point << ':' << low
            << ':' << high << "|records="
            << names(result.first, result.second) << '\n';
}

void dumpWalk(const string &label, const TestMap &map)
{
  std::cout << "case=" << label << "|walk="
            << names(map.begin(), map.end()) << "|list=" << listNames(map)
            << '\n';
}

void dumpOverlap(const string &label, const TestMap &map,
                 TestRecord::linetype point, TestRecord::linetype end)
{
  TestMap::const_iterator iter = map.find_overlap(point, end);
  std::cout << "case=" << label << "|overlap=" << point << ':' << end
            << "|record=";
  if (iter == map.end()) std::cout << "null";
  else std::cout << (*iter).getName();
  std::cout << '\n';
}

void dumpCursor(const string &label, const TestMap &map,
                TestMap::const_iterator cursor)
{
  std::cout << "case=" << label << "|cursor=";
  if (cursor == map.end()) std::cout << "end";
  else std::cout << (*cursor).getName();
  std::cout << '\n';
}

void equalRangeCases(void)
{
  TestMap map;
  map.insert(TestRecord::InitData("eq_a", 5), 10, 20);
  dumpWalk("equal_after_a", map);
  map.insert(TestRecord::InitData("eq_b", 5), 10, 20);
  dumpWalk("equal_after_b", map);
  map.insert(TestRecord::InitData("sub_hi", 9), 10, 20);
  dumpWalk("equal_after_hi", map);
  map.insert(TestRecord::InitData("sub_lo", 1), 10, 20);
  dumpWalk("equal_after_lo", map);
  dumpFind("equal_range", map, 15);
  dumpFindSubsort("equal_subsort_window", map, 15, 5, 5);
  dumpWalk("equal_ordered_walk", map);
}

void wideNarrowCases(void)
{
  TestMap wideFirst;
  wideFirst.insert(TestRecord::InitData("wide", 5), 100, 140);
  wideFirst.insert(TestRecord::InitData("narrow", 2), 110, 120);
  dumpFind("wide_first_left", wideFirst, 105);
  dumpFind("wide_first_middle", wideFirst, 115);
  dumpFind("wide_first_right", wideFirst, 130);
  dumpWalk("wide_first_walk", wideFirst);

  TestMap narrowFirst;
  narrowFirst.insert(TestRecord::InitData("narrow", 2), 110, 120);
  narrowFirst.insert(TestRecord::InitData("wide", 5), 100, 140);
  dumpFind("narrow_first_left", narrowFirst, 105);
  dumpFind("narrow_first_middle", narrowFirst, 115);
  dumpFind("narrow_first_right", narrowFirst, 130);
  dumpWalk("narrow_first_walk", narrowFirst);
}

void splitEraseCases(void)
{
  TestMap map;
  map.insert(TestRecord::InitData("outer", 5), 200, 240);
  std::list<TestRecord>::iterator inner = map.insert(
      TestRecord::InitData("inner", 3), 210, 230);
  std::list<TestRecord>::iterator right = map.insert(
      TestRecord::InitData("right", 7), 220, 250);
  dumpWalk("split_before_erase", map);
  dumpFind("split_three_way", map, 225);

  map.erase(inner);
  dumpWalk("erase_inner_zip_left", map);
  dumpFind("erase_inner_middle", map, 215);
  dumpFind("erase_inner_overlap", map, 225);

  map.erase(right);
  dumpWalk("erase_right_zip_both", map);
  dumpFind("erase_right_outer", map, 225);
}

void overlapAndRangeWalkCases(void)
{
  TestMap map;
  map.insert(TestRecord::InitData("left", 4), 300, 305);
  map.insert(TestRecord::InitData("right_hi", 8), 310, 320);
  map.insert(TestRecord::InitData("right_lo", 2), 310, 315);
  dumpOverlap("gap_hit", map, 306, 312);
  dumpOverlap("gap_miss", map, 306, 309);
  dumpOverlap("inside_order", map, 312, 312);
  std::cout << "case=bounded_walk|range=304:316|records="
            << names(map.find_begin(304), map.find_end(316)) << '\n';
}

void equivalentSplitCases(void)
{
  TestMap map;
  map.insert(TestRecord::InitData("same_a", 5), 400, 440);
  std::list<TestRecord>::iterator sameB = map.insert(
      TestRecord::InitData("same_b", 5), 400, 440);
  std::list<TestRecord>::iterator cut = map.insert(
      TestRecord::InitData("cut", 5), 410, 420);
  dumpWalk("equivalent_split", map);
  dumpFind("equivalent_split_middle", map, 415);
  map.erase(cut);
  dumpWalk("equivalent_erase_cut", map);
  map.erase(sameB);
  dumpWalk("equivalent_erase_peer", map);
}

void stableCursorCases(void)
{
  TestMap before;
  before.insert(TestRecord::InitData("A", 5), 100, 110);
  TestMap::const_iterator beforeCursor = before.find_begin(100);
  dumpCursor("cursor_insert_before_initial", before, beforeCursor);
  before.insert(TestRecord::InitData("B", 5), 50, 60);
  dumpCursor("cursor_insert_before_after", before, beforeCursor);
  before.erase(beforeCursor);
  dumpWalk("cursor_insert_before_erase", before);

  TestMap after;
  after.insert(TestRecord::InitData("A", 5), 100, 110);
  TestMap::const_iterator afterCursor = after.find_begin(100);
  after.insert(TestRecord::InitData("C", 5), 150, 160);
  dumpCursor("cursor_insert_after", after, afterCursor);
  after.erase(afterCursor);
  dumpWalk("cursor_insert_after_erase", after);

  TestMap equivalent;
  equivalent.insert(TestRecord::InitData("eq_a", 5), 10, 20);
  TestMap::const_iterator equivalentCursor = equivalent.find_begin(15);
  equivalent.insert(TestRecord::InitData("eq_b", 5), 10, 20);
  dumpCursor("cursor_equivalent_after", equivalent, equivalentCursor);
  equivalent.erase(equivalentCursor);
  dumpWalk("cursor_equivalent_erase", equivalent);

  TestMap split;
  split.insert(TestRecord::InitData("wide", 5), 100, 140);
  TestMap::const_iterator splitCursor = split.find_begin(120);
  split.insert(TestRecord::InitData("narrow", 2), 110, 120);
  dumpCursor("cursor_unzip_after", split, splitCursor);
  split.erase(splitCursor);
  dumpWalk("cursor_unzip_erase", split);

  TestMap endMap;
  TestMap::const_iterator oldEnd = endMap.end();
  endMap.insert(TestRecord::InitData("only", 5), 1, 2);
  std::cout << "case=cursor_end_insert|same="
            << (oldEnd == endMap.end() ? 1 : 0) << '\n';
}

void cursorInvalidationMatrix(void)
{
  TestMap map;
  map.insert(TestRecord::InitData("outer", 5), 200, 240);
  std::list<TestRecord>::iterator innerRecord = map.insert(
      TestRecord::InitData("inner", 3), 210, 230);
  TestMap::const_iterator leftOuter = map.find_begin(205);
  std::pair<TestMap::const_iterator, TestMap::const_iterator> middle =
      map.find(220);
  TestMap::const_iterator innerMiddle = middle.first;
  TestMap::const_iterator middleOuter = innerMiddle;
  ++middleOuter;
  TestMap::const_iterator rightOuter = map.find_begin(235);
  TestMap::const_iterator oldEnd = map.end();

  std::cout << "case=cursor_matrix_before|left=" << (*leftOuter).getName()
            << "|inner=" << (*innerMiddle).getName()
            << "|middle=" << (*middleOuter).getName()
            << "|right=" << (*rightOuter).getName() << '\n';
  map.erase(innerRecord);
  // std::multiset::erase invalidates only erased elements.  erase(inner)
  // removes innerMiddle directly; zip(209) erases leftOuter; zip(230)
  // erases middleOuter; rightOuter is only extended in place and survives.
  std::cout << "case=cursor_matrix_after|left=invalid|inner=invalid"
            << "|middle=invalid|right=" << (*rightOuter).getName()
            << "|end_same=" << (oldEnd == map.end() ? 1 : 0) << '\n';
  map.erase(rightOuter);
  dumpWalk("cursor_matrix_right_erase", map);
}

} // anonymous namespace

int main(void)
{
  equalRangeCases();
  wideNarrowCases();
  splitEraseCases();
  overlapAndRangeWalkCases();
  equivalentSplitCases();
  stableCursorCases();
  cursorInvalidationMatrix();
  return 0;
}

File: tests/oracle/rangemap_common_refinement_1204.rs
// Locked Ghidra 12.0.4 rangemap<> oracle for
// RANGEMAP-COMMON-REFINEMENT-0001 (Rust comparand).

use rugra::rangemap::{RangeMap, RangeMapIter, RangeRecord, RangeSubsort};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
struct Subsort(i32);

impl RangeSubsort for Subsort {
    fn minimum() -> Self {
        Self(0)
    }

    fn maximum() -> Self {
        Self(1_000_000)
    }
}

#[derive(Debug)]
struct TestRecord {
    name: &'static str,
    first: u64,
    last: u64,
    subsort: Subsort,
}

impl TestRecord {
    fn new(name: &'static str, subsort: i32, first: u64, last: u64) -> Self {
        Self {
            name,
            first,
            last,
            subsort: Subsort(subsort),
        }
    }
}

impl RangeRecord for TestRecord {
    type Subsort = Subsort;

    fn first(&self) -> u64 {
        self.first
    }

    fn last(&self) -> u64 {
        self.last
    }

    fn subsort(&self) -> Self::Subsort {
        self.subsort.clone()
    }
}

fn names<'a>(records: impl Iterator<Item = &'a TestRecord>) -> String {
    records
        .map(|record| record.name)
        .collect::<Vec<_>>()
        .join(",")
}

fn dump_find(label: &str, map: &RangeMap<TestRecord>, point: u64) {
    println!(
        "case={label}|find={point}|records={}",
        names(map.find(point))
    );
}

fn dump_find_subsort(label: &str, map: &RangeMap<TestRecord>, point: u64, low: i32, high: i32) {
    println!(
        "case={label}|find_sub={point}:{low}:{high}|records={}",
        names(map.find_with_subsort(point, &Subsort(low), &Subsort(high)))
    );
}

fn dump_walk(label: &str, map: &RangeMap<TestRecord>) {
    println!(
        "case={label}|walk={}|list={}",
        names(map.iter()),
        names(map.records())
    );
}

fn dump_overlap(label: &str, map: &RangeMap<TestRecord>, point: u64, end: u64) {
    let record = map
        .find_overlap(point, end)
        .map_or("null", |record| record.name);
    println!("case={label}|overlap={point}:{end}|record={record}");
}

fn dump_cursor(label: &str, map: &RangeMap<TestRecord>, cursor: rugra::rangemap::RangeMapCursor) {
    let record = map
        .record_at_cursor(cursor)
        .map_or("end", |record| record.name);
    println!("case={label}|cursor={record}");
}

fn equal_range_cases() {
    let mut map = RangeMap::new();
    map.insert(TestRecord::new("eq_a", 5, 10, 20));
    dump_walk("equal_after_a", &map);
    map.insert(TestRecord::new("eq_b", 5, 10, 20));
    dump_walk("equal_after_b", &map);
    map.insert(TestRecord::new("sub_hi", 9, 10, 20));
    dump_walk("equal_after_hi", &map);
    map.insert(TestRecord::new("sub_lo", 1, 10, 20));
    dump_walk("equal_after_lo", &map);
    dump_find("equal_range", &map, 15);
    dump_find_subsort("equal_subsort_window", &map, 15, 5, 5);
    dump_walk("equal_ordered_walk", &map);
}

fn wide_narrow_cases() {
    let mut wide_first = RangeMap::new();
    wide_first.insert(TestRecord::new("wide", 5, 100, 140));
    wide_first.insert(TestRecord::new("narrow", 2, 110, 120));
    dump_find("wide_first_left", &wide_first, 105);
    dump_find("wide_first_middle", &wide_first, 115);
    dump_find("wide_first_right", &wide_first, 130);
    dump_walk("wide_first_walk", &wide_first);

    let mut narrow_first = RangeMap::new();
    narrow_first.insert(TestRecord::new("narrow", 2, 110, 120));
    narrow_first.insert(TestRecord::new("wide", 5, 100, 140));
    dump_find("narrow_first_left", &narrow_first, 105);
    dump_find("narrow_first_middle", &narrow_first, 115);
    dump_find("narrow_first_right", &narrow_first, 130);
    dump_walk("narrow_first_walk", &narrow_first);
}

fn split_erase_cases() {
    let mut map = RangeMap::new();
    map.insert(TestRecord::new("outer", 5, 200, 240));
    let inner = map.insert(TestRecord::new("inner", 3, 210, 230));
    let right = map.insert(TestRecord::new("right", 7, 220, 250));
    dump_walk("split_before_erase", &map);
    dump_find("split_three_way", &map, 225);

    map.erase(inner);
    dump_walk("erase_inner_zip_left", &map);
    dump_find("erase_inner_middle", &map, 215);
    dump_find("erase_inner_overlap", &map, 225);

    map.erase(right);
    dump_walk("erase_right_zip_both", &map);
    dump_find("erase_right_outer", &map, 225);
}

fn overlap_and_range_walk_cases() {
    let mut map = RangeMap::new();
    map.insert(TestRecord::new("left", 4, 300, 305));
    map.insert(TestRecord::new("right_hi", 8, 310, 320));
    map.insert(TestRecord::new("right_lo", 2, 310, 315));
    dump_overlap("gap_hit", &map, 306, 312);
    dump_overlap("gap_miss", &map, 306, 309);
    dump_overlap("inside_order", &map, 312, 312);
    let begin = map.find_begin(304);
    let end = map.find_end(316);
    let bounded: RangeMapIter<'_, TestRecord> = map.iter_between(begin, end);
    println!("case=bounded_walk|range=304:316|records={}", names(bounded));
}

fn equivalent_split_cases() {
    let mut map = RangeMap::new();
    map.insert(TestRecord::new("same_a", 5, 400, 440));
    let same_b = map.insert(TestRecord::new("same_b", 5, 400, 440));
    let cut = map.insert(TestRecord::new("cut", 5, 410, 420));
    dump_walk("equivalent_split", &map);
    dump_find("equivalent_split_middle", &map, 415);
    map.erase(cut);
    dump_walk("equivalent_erase_cut", &map);
    map.erase(same_b);
    dump_walk("equivalent_erase_peer", &map);
}

fn stable_cursor_cases() {
    let mut before = RangeMap::new();
    before.insert(TestRecord::new("A", 5, 100, 110));
    let before_cursor = before.find_begin(100);
    dump_cursor("cursor_insert_before_initial", &before, before_cursor);
    before.insert(TestRecord::new("B", 5, 50, 60));
    dump_cursor("cursor_insert_before_after", &before, before_cursor);
    before.erase_at(before_cursor);
    dump_walk("cursor_insert_before_erase", &before);

    let mut after = RangeMap::new();
    after.insert(TestRecord::new("A", 5, 100, 110));
    let after_cursor = after.find_begin(100);
    after.insert(TestRecord::new("C", 5, 150, 160));
    dump_cursor("cursor_insert_after", &after, after_cursor);
    after.erase_at(after_cursor);
    dump_walk("cursor_insert_after_erase", &after);

    let mut equivalent = RangeMap::new();
    equivalent.insert(TestRecord::new("eq_a", 5, 10, 20));
    let equivalent_cursor = equivalent.find_begin(15);
    equivalent.insert(TestRecord::new("eq_b", 5, 10, 20));
    dump_cursor("cursor_equivalent_after", &equivalent, equivalent_cursor);
    equivalent.erase_at(equivalent_cursor);
    dump_walk("cursor_equivalent_erase", &equivalent);

    let mut split = RangeMap::new();
    split.insert(TestRecord::new("wide", 5, 100, 140));
    let split_cursor = split.find_begin(120);
    split.insert(TestRecord::new("narrow", 2, 110, 120));
    dump_cursor("cursor_unzip_after", &split, split_cursor);
    split.erase_at(split_cursor);
    dump_walk("cursor_unzip_erase", &split);

    let mut end_map = RangeMap::new();
    let old_end = end_map.find_begin(0);
    end_map.insert(TestRecord::new("only", 5, 1, 2));
    println!(
        "case=cursor_end_insert|same={}",
        if end_map.cursor_is_valid(old_end) && end_map.record_at_cursor(old_end).is_none() {
            1
        } else {
            0
        }
    );
}

fn cursor_invalidation_matrix() {
    let mut map = RangeMap::new();
    map.insert(TestRecord::new("outer", 5, 200, 240));
    let inner_id = map.insert(TestRecord::new("inner", 3, 210, 230));
    let left_outer = map.find_begin(205);
    let inner_middle = map.find_begin(220);
    let middle_outer = map.next_cursor(inner_middle).unwrap();
    let right_outer = map.find_begin(235);
    let old_end = map.find_end(u64::MAX);

    println!(
        "case=cursor_matrix_before|left={}|inner={}|middle={}|right={}",
        map.record_at_cursor(left_outer).unwrap().name,
        map.record_at_cursor(inner_middle).unwrap().name,
        map.record_at_cursor(middle_outer).unwrap().name,
        map.record_at_cursor(right_outer).unwrap().name
    );
    map.erase(inner_id);
    println!(
        "case=cursor_matrix_after|left={}|inner={}|middle={}|right={}|end_same={}",
        if map.cursor_is_valid(left_outer) {
            "valid"
        } else {
            "invalid"
        },
        if map.cursor_is_valid(inner_middle) {
            "valid"
        } else {
            "invalid"
        },
        if map.cursor_is_valid(middle_outer) {
            "valid"
        } else {
            "invalid"
        },
        map.record_at_cursor(right_outer).unwrap().name,
        if map.cursor_is_valid(old_end) && map.record_at_cursor(old_end).is_none() {
            1
        } else {
            0
        }
    );
    map.erase_at(right_outer);
    dump_walk("cursor_matrix_right_erase", &map);
}

fn main() {
    equal_range_cases();
    wide_narrow_cases();
    split_erase_cases();
    overlap_and_range_walk_cases();
    equivalent_split_cases();
    stable_cursor_cases();
    cursor_invalidation_matrix();
}
